QEMU-level migration parallelism (a.k.a. multifd) + Upgrade QEMU to 7.2.0-14.el9

iholder101 opened the pull request:
**What this PR does / why we need it**:
This PR brings initial preview support to QEMU-level parallel migrations (a.k.a. multifd migrations), that is, using multiple parallel threads in order to migrate. This is achieved by using `VIR_MIGRATE_PARALLEL` migration parameter.
From [the docs](https://libvirt.org/html/libvirt-libvirt-domain.html):
```
Send memory pages to the destination host through several network connections.
```

With just a single migration connection and sufficient network bandwidth, the single migration thread can become a bottleneck due to a limited single-CPU capacity. By using multiple connections fed with multiple threads run by multiple CPUs, a full network capacity can be utilized. For instance, 8-16 connections can saturate 100 Gbps network bandwidth.

Additionally, multifd implementation of data transfer is simpler and more efficient than the traditional migration mechanism and can be faster even with lower network bandwidths. This concerns only the way memory data is transferred, the other migration concepts such as counting iterations or setting the maximum downtime are unaffected.

This feature is especially valuable for environments where VMs have a high dirty-rate, but network bandwidth is high.

**How to enable**:
As a start, this feature can be enabled on a per-VM/VMI basis via an annotation:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  annotations:
    "kubevirt.io/multiThreadedQemuMigration": "5"
```

Bear in mind that this is an initial dev-preview, therefore no API changes are introduced. In the future such APIs would be introduced, probably at the Kubevirt CR / migration policy level.

<s>
Migration Policies are design follows the following principle: configurations can be enabled by both Kubevirt CR (which serves as a cluster-wide default) and Migration Policies, while the latter has precedence. This feature follows that design.

To enable this as a cluster-wide configuration, edit Kubevirt CR as follows:
```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  configuration:
    migrations:
      parallelMigrationThreads: 123
```

To enable as part of a migration policy:
```yaml
apiVersion: migrations.kubevirt.io/v1alpha1
kind: MigrationPolicy
metadata:
  name: my-awesome-policy
spec:
  parallelMigrationThreads: 123
```

As mentioned above, migration policies always have precedence over Kubevirt CR.
</s>

**Special notes for your reviewer**:
<s>Currently blocked because bug in QEMU: https://bugzilla.redhat.com/show_bug.cgi?id=2137740, https://bugzilla.redhat.com/show_bug.cgi?id=2169732</s>

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
QEMU-level migration parallelism (a.k.a. multifd) + Upgrade QEMU to 7.2.0-11.el9
```

iholder-redhat commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-compute
/test pull-kubevirt-e2e-k8s-1.23-sig-compute

vladikr commented:
Great to see this PR, @iholder-redhat 

One thing to keep in mind is that post-copy migration is not supported with parallel migration. so we will need some handling around that.

iholder101 commented:
Currently blocked because bug in QEMU: https://bugzilla.redhat.com/show_bug.cgi?id=2137740

iholder101 commented:
/remove-lifecycle stale

iholder101 commented:
/cc @vladikr

iholder101 commented:
/retest-required

vasiliy-ul commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -90,24 +90,27 @@ type inflightMigrationAborted struct {
 	if migratePaused {
 		migrateFlags |= libvirt.MIGRATE_PAUSED
 	}
+	if options.ParallelMigrationThreads != nil {
```
Are `0` and `1` valid values? E.g. what happens if `ParallelMigrationThreads` is set to `0`? Also, should we enable the parallel migration flag if `ParallelMigrationThreads` is `1`?

iholder101 replied:
Thanks a lot for bringing this up!
Can you PTAL on my last commit? It rejects Kubevirt CR / migration policies with `ParallelMigrationThreads <= 1`.

p.s. - For every migration config, we basically need to perform very similar check when it comes to Kubevirt CR and migration policies. As a follow-up PR, we should somehow converge this logic into one place so it will both be checked in the same way for the CR and the policies, and the code won't be duplicated.

Thanks a lot!

vladikr commented:
> > Great to see this PR, @iholder-redhat
> > One thing to keep in mind is that post-copy migration is not supported with parallel migration. so we will need some handling around that.
> 
> From my understanding this actually should be supported: https://bugzilla.redhat.com/show_bug.cgi?id=2106726. But TBH I'm not really sure what the logic is on that scenario.

My understanding is that parallel migration is not supported with post-copy, but I might be wrong.
In any case, it's easy to check. ;)

@dagrh would surely know. 

> 
> Do you think we should disallow that? In your opinion should the migration fail? or simply not use parallel migrations if post-copy is being used?

vladikr commented:
It's great to see that we are making progress on this, however, I'd be careful with just providing a way to enable this as it is suggested right now.

I have a few concerns.

 - There should be a limit to the number of threads configured per migration. I'm not sure what's the optimal number, but it should depend on the available resources - I don't think that the admin should just set a number.
 - parallel migration is also CPU intensive. In some scenarios, it just doesn't make sense to enable this.
   - In the dedicated CPUs case, these are the workloads that require predictable/lower latency (probably also have a high dirty rate), enabling parallel migration would increase the CPU load in the compute container which will effectively increase the latency and will compete with vcpus for the cpu time...
   - In the isolatedEmulatorThread case, all of these additional threads should be moved to the housekeeping cgroup and all of these threads will use a single core...

My suggestion is that we should make kubevirt smart enough to choose a correct migration policy that enables parallel migration and set the number of threads dynamically according to some heuristic.

I think that as a first step, to being able to experiment with parallel migrations and build a better understanding around it, we should enable this as an experimental feature without making any changes.
We could simply introduce a feature gate and annotation on the VMI that will specify the desired amount of threads - virt-handler then can enabled this.

iholder101 commented:
> It's great to see that we are making progress on this, however, I'd be careful with just providing a way to enable this as it is suggested right now.
> 
> I have a few concerns.
> 
>  * There should be a limit to the number of threads configured per migration. I'm not sure what's the optimal number, but it should depend on the available resources - I don't think that the admin should just set a number.

I agree that the number of threads should be limited.

> * parallel migration is also CPU intensive. In some scenarios, it just doesn't make sense to enable this.      
>   * In the dedicated CPUs case, these are the workloads that require predictable/lower latency (probably also have a high dirty rate), enabling parallel migration would increase the CPU load in the compute container which will effectively increase the latency and will compete with vcpus for the cpu time...

As you know this is going to change in the future.
Off-topic: I never understood why we don't use the housekeeping cgroup in this case as well.

>   * In the isolatedEmulatorThread case, all of these additional threads should be moved to the housekeeping cgroup and all of these threads will use a single core...

To be fair, we're talking about a full **dedicated core** that serves libvirtd, virt-launcher, and some other processes that aren't expected to be very CPU-bound.

> My suggestion is that we should make kubevirt smart enough to choose a correct migration policy that enables parallel migration and set the number of threads dynamically according to some heuristic.

I agree that this should be the vision forward

> I think that as a first step, to being able to experiment with parallel migrations and build a better understanding around it, we should enable this as an experimental feature without making any changes. We could simply introduce a feature gate and annotation on the VMI that will specify the desired amount of threads - virt-handler then can enabled this.

How about using an annotation on a migration policy level instead of a VMI level?
The reason I think this is better is that it would be much easier to test and explore this way since we can easily apply new settings for a group of VMs, that could be already running, fully transparently.

WDYT?

iholder101 commented:
Updated PR:
* All new APIs are now removed. The thread count is now controlled via an annotation.
* I didn't restrict the max number of threads. Since this is just a tech review, let's take the time to explore what a reasonable max could be.
* I've opened (and closed) this PR: https://github.com/kubevirt/kubevirt/pull/9512. It's just for reference purposes for the day we would want to introduce the API.

PTAL

iholder101 commented:
> > > > Great to see this PR, @iholder-redhat
> > > > One thing to keep in mind is that post-copy migration is not supported with parallel migration. so we will need some handling around that.
> > > 
> > > 
> > > From my understanding this actually should be supported: https://bugzilla.redhat.com/show_bug.cgi?id=2106726. But TBH I'm not really sure what the logic is on that scenario.
> > 
> > 
> > My understanding is that parallel migration is not supported with post-copy, but I might be wrong. In any case, it's easy to check. ;)
> > @dagrh would surely know.
> > > Do you think we should disallow that? In your opinion should the migration fail? or simply not use parallel migrations if post-copy is being used?
> 
> We probably need to check with Leonardo Bras (sorry not sure of his Github); we were looking to disable it in RHEL8 because it was unstable, but I know Leo has been fixing a bunch of things in it, so not sure what the rhel9 or upstream status currently is. I'd certainly say the postcopy+multifd is a bit bleeding edge.

Thanks a lot @dagrh!
@vladikr - are you okay with not restricting it at this tech view point?

iholder101 commented:
@vladikr FYI - updated QEMU to be affected by the changes.

/cc @andreabolognani 
Thanks a lot for your help!

dagrh commented:
@iholder101 FYI Leo has just posted a patch to disable multifd+postcopy upstream in qemu until it shakes out a bit; so I would stay away from that combo for now.

iholder101 commented:
> @iholder101 FYI Leo has just posted a patch to disable multifd+postcopy upstream in qemu until it shakes out a bit; so I would stay away from that combo for now.

Thank you for bringing it up @dagrh! Appreciate it.
@vladikr: Since this is a dev preview with no API, I suggest to defer enforcing no post-copy+multifd to future PRs. WDYT?

andreabolognani commented:
@iholder101 looking just at the package update part, because the rest is going way over my head :)

I think you should either update all virtualization packages, including libvirt, or limit the update to QEMU, which is the package that contains the fix you care about. This middle ground where you update just a subset of the packages doesn't feel right.

I suggest that you just drop the commit that updates passt and only keep the QEMU update in. That way, you also reduce the risk of unrelated changes possibly getting in the way of having your feature merged. I'll take care of updating libvirt, passt and whatever other packages might need it myself in a week or so.

xpivarc commented on WORKSPACE:
```diff
@@ -663,57 +663,39 @@ rpm(
-    sha256 = "2119214b895edbb45942003d7bca5c3ed9d43061c75a9804c0ffb12cd041878e",
-    urls = [
-        "http://mirror.stream.centos.org/9-stream/BaseOS/aarch64/os/Packages/centos-gpg-keys-9.0-19.el9.noarch.rpm",
-        "https://storage.googleapis.com/builddeps/2119214b895edbb45942003d7bca5c3ed9d43061c75a9804c0ffb12cd041878e",
```
We have automation which takes the rpms and will mirror these to the google bucket.

jean-edouard reviewed (commented):
/approve
/hold
Please just fix the PR description to reflect the switch from API to annotation.
I would also remove the word "Support" from the PR title, since this doesn't sound like a supported feature just yet.
Thank you!

iholder101 commented:
> /approve /hold Please just fix the PR description to reflect the switch from API to annotation. I would also remove the word "Support" from the PR title, since this doesn't sound like a supported feature just yet. Thank you!

Done!
Thank you!

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.25-sig-compute

vladikr commented:
Thanks @iholder101 !
Looks good to me.
/lgtm

iholder101 commented:
/unhold

vladikr commented:
/test pull-kubevirt-e2e-k8s-1.25-sig-compute-migrations

vladikr commented:
/lgtm
